### PR TITLE
Make `BalanceDelta` great

### PR DESCRIPTION
**Pull request by shuhuiluo**
## Related Issue
Which issue does this pull request resolve?

## Description of changes
Optimize `eq`, `neq`, `add` and `sub` for `BalanceDelta`. Refactor `amount0` and `amount1` for correctness.

### CODE REVIEW

File: src/types/BalanceDelta.sol [hunk 1]
```diff
@@ -1,48 +1,67 @@
...
+    int256 res1;
+    assembly {
+        let a0 := sar(128, a)
+        let a1 := signextend(15, a)
+        let b0 := sar(128, b)
+        let b1 := signextend(15, b)
+        res0 := add(a0, b0)
+        res1 := add(a1, b1)
+    }
+    return toBalanceDelta(SafeCast.toInt128(res0), SafeCast.toInt128(res1));
```
Comment (hensha256) [inline src/types/BalanceDelta.sol:28]: I have a slight preference for `res0.toInt128()`

File: src/types/BalanceDelta.sol [hunk 2]
```diff
@@ -1,48 +1,67 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.20;
 
+import {SafeCast} from "../libraries/SafeCast.sol";
+
 type BalanceDelta is int256;
```
Comment (hensha256) [inline src/types/BalanceDelta.sol:6]: would you mind putting a comment above BalanceDelta describing that BalanceDelta is 2 int128s packed together, one for 0 and one for 1?
I just think with these assembly changes, its becoming less obvious at a quick glance what this type is haha

**Review by hensha256 — COMMENTED** [review]
generally love it, just a couple of requests!

File: src/types/BalanceDelta.sol (hunk 1 above)
Comment (shuhuiluo) [reply src/types/BalanceDelta.sol]: Done.

File: src/types/BalanceDelta.sol (hunk 2 above)
Comment (shuhuiluo) [reply src/types/BalanceDelta.sol:8]: Added.

**Review by hensha256 — APPROVED** [review]

**Pull request merged**